tests/unit/test_main.py

```diff
@@ -58,43 +74,575 @@ def test_verify_api_urls(self):
+        """Helper method to patch controller to return switch/flow data."""
+        # patching the flow_stats object in the switch
+        self.napp.controller.switches = MagicMock()
+        flow = self._get_mocked_flow_stats()
+        flow.id = flow_id
+        switch = MagicMock()
+        switch.generic_flows = [flow]
+        self.napp.controller.switches.values.return_value = [switch]
+        self.napp.controller.get_switch_by_dpid = MagicMock()
+        self.napp.controller.get_switch_by_dpid.return_value = switch
+
+    def _get_rest_response(self, rest_name, url_id):
+        """Helper method to call a rest endpoint."""
+        # call rest
+        api = get_test_client(get_controller_mock(), self.napp)
+        url = f"{self.server_name_url}/{rest_name}/{url_id}"
+        response = api.get(url, content_type="application/json")
+
+        return response
+
+    # pylint: disable=no-self-use
+    def _get_mocked_flow_stats(self):
+        """Helper method to create a mock flow_stats object."""
+        flow_stats = MagicMock()
+        flow_stats.id = 123
+        flow_stats.byte_count = 10
+        flow_stats.duration_sec = 20
+        flow_stats.duration_nsec = 30
+        flow_stats.packet_count = 40
+        return flow_stats
+
+    def _get_mocked_flow_base(self):
+        """Helper method to create a mock flow object."""
+        flow = MagicMock()
+        flow.id = 456
+        flow.switch = None
+        flow.table_id = None
+        flow.match = None
+        flow.priority = None
+        flow.idle_timeout = None
+        flow.hard_timeout = None
+        flow.cookie = None
+        flow.stats = self._get_mocked_flow_stats()
+        return flow
+
+    @patch("napps.amlight.flow_stats.main.GenericFlow.from_flow_stats")
+    def test_handle_stats_reply(self, mock_from_flow):
+        """Test handle_stats_reply rest call."""
+        mock_from_flow.return_value = self._get_mocked_flow_base()
+
+        def side_effect(event):
+            self.assertTrue(f"{event}", "amlight/flow_stats.flows_updated")
+            self.assertTrue(event.content["switch"], 111)
+
+        self.napp.controller = MagicMock()
+        self.napp.controller.buffers.app.put.side_effect = side_effect
+
+        msg = MagicMock()
+        msg.flags.value = 2
+        msg.body = [self._get_mocked_flow_stats()]
+        event_switch = MagicMock()
+        event_switch.generic_flows = []
+        event_switch.dpid = 111
+        self.napp.handle_stats_reply(msg, event_switch)
+        time.sleep(0.5)
+
+        # Check if important trace dont trigger the event
+        # It means that the CP trace is the same to the DP trace
+        self.napp.controller.buffers.app.put.assert_called_once()
+
+        # Check mocked flow id
+        self.assertEqual(event_switch.generic_flows[0].id, 456)
+
+    @patch("napps.amlight.flow_stats.main.Main.handle_stats_reply")
+    def test_handle_stats_reply_0x01(self, mock_handle_stats):
```

Comment:
@rmotitsuki, thanks for enhancing test cov for some core parts. Let's get rid of these `time.sleep` in unit tests, they will lead to slowness and might contribute to test flakiness, for threaded `listen_to` handlers we've been only testing out the inner method instead, and then intentionally leaving the handler to only call the inner method and minimizing conditionals there too, could you apply this patch or refactor to something similar?

```
diff --git a/tests/unit/test_main.py b/tests/unit/test_main.py
index d8ebdb3..dd28bcc 100644
--- a/tests/unit/test_main.py
+++ b/tests/unit/test_main.py
@@ -344,7 +344,6 @@ class TestMain(TestCase):
         event_switch.generic_flows = []
         event_switch.dpid = 111
         self.napp.handle_stats_reply(msg, event_switch)
-        time.sleep(0.5)
 
         # Check if important trace dont trigger the event
         # It means that the CP trace is the same to the DP trace
@@ -359,76 +358,20 @@ class TestMain(TestCase):
         flow_msg = MagicMock()
         flow_msg.body = "A"
         flow_msg.body_type = StatsType.OFPST_FLOW
-
         switch_v0x01 = get_switch_mock("00:00:00:00:00:00:00:01", 0x01)
-
-        name = "kytos/of_core.v0x01.messages.in.ofpt_stats_reply"
-        content = {"source": switch_v0x01.connection, "message": flow_msg}
-        event = get_kytos_event_mock(name=name, content=content)
-        self.napp.handle_stats_reply_0x01(event)
-        # Event listener is async. Just waiting for the Mock to return
-        time.sleep(0.5)
+        self.napp.handle_stats_reply(flow_msg, switch_v0x01)
         mock_handle_stats.assert_called_once()
 
-    @patch("napps.amlight.flow_stats.main.Main.handle_stats_reply")
-    def test_handle_stats_reply_0x01__fail(self, mock_handle_stats):
-        """Test handle stats reply."""
-        flow_msg = MagicMock()
-        flow_msg.body = "A"
-        flow_msg.body_type = StatsType.OFPST_DESC
-
-        switch_v0x01 = get_switch_mock("00:00:00:00:00:00:00:01", 0x01)
-
-        name = "kytos/of_core.v0x01.messages.in.ofpt_stats_reply"
-        content = {"source": switch_v0x01.connection, "message": flow_msg}
-        event = get_kytos_event_mock(name=name, content=content)
-        self.napp.handle_stats_reply_0x01(event)
-        # Event listener is async. Just waiting for the Mock to return
-        time.sleep(0.5)
-        mock_handle_stats.assert_not_called()
-
     @patch("napps.amlight.flow_stats.main.Main.handle_stats_reply")
     def test_handle_stats_reply_0x04(self, mock_handle_stats):
         """Test handle stats reply."""
         flow_msg = MagicMock()
         flow_msg.body = "A"
         flow_msg.multipart_type = MultipartType.OFPMP_FLOW
-
         switch_v0x04 = get_switch_mock("00:00:00:00:00:00:00:01", 0x04)
-
-        name = "kytos/of_core.v0x04.messages.in.ofpt_multipart_reply"
-        content = {"source": switch_v0x04.connection, "message": flow_msg}
-        event = get_kytos_event_mock(name=name, content=content)
-        event.content["message"] = flow_msg
-
-        self.napp.handle_stats_reply_0x04(event)
-        # Event listener is async. Just waiting for the Mock to return
-        time.sleep(0.5)
-
+        self.napp.handle_stats_reply(flow_msg, switch_v0x04)
         mock_handle_stats.assert_called_once()
 
-    @patch("napps.amlight.flow_stats.main.Main.handle_stats_reply")
-    def test_handle_stats_reply_0x04_fail(self, mock_handle_stats):
-        """Test handle stats reply."""
-        flow_msg = MagicMock()
-        flow_msg.body = "A"
-
-        flow_msg.multipart_type = MultipartType.OFPMP_PORT_DESC
-
-        switch_v0x04 = get_switch_mock("00:00:00:00:00:00:00:01", 0x04)
-
-        name = "kytos/of_core.v0x04.messages.in.ofpt_multipart_reply"
-        content = {"source": switch_v0x04.connection, "message": flow_msg}
-
-        event = get_kytos_event_mock(name=name, content=content)
-        event.content["message"] = flow_msg
-
-        self.napp.handle_stats_reply_0x04(event)
-        # Event listener is async. Just waiting for the Mock to return
-        time.sleep(0.5)
-
-        mock_handle_stats.assert_not_called()
-
 
 # pylint: disable=too-many-public-methods, too-many-lines
 class TestGenericFlow(TestCase):
```